File: csharp/dnd-character/DndCharacter.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class DndCharacter
{
    public int Strength { get; private set; }
    public int Dexterity { get; private set;}
    public int Constitution { get; private set;}
    public int Intelligence { get; private set;}
    public int Wisdom { get; private set;}
    public int Charisma { get; private set;}
    public int Hitpoints { get; private set;}

    public static int Modifier(int score)
    {
        return (int)Math.Floor((score-10)/2.0);
    }

    public static int Ability() 
    {
        List<int> rollList = new List<int>();
        Random r = new Random();
        for (int i=0; i<4;i++)
            rollList.Add(r.Next(1, 7));
        rollList = rollList.OrderByDescending(s => s).ToList<int>();
        return rollList.Take(3).Sum();
    }

    public static DndCharacter Generate()
    {
        DndCharacter character = new DndCharacter();
        character.Charisma = Ability();
        character.Strength = Ability();
        character.Dexterity = Ability();
        character.Constitution = Ability();
        character.Wisdom = Ability();
        character.Intelligence = Ability();
        character.Hitpoints = 10 + Modifier(character.Constitution);
        return character;
    }
}

File: java/dnd-character/src/main/java/DnDCharacter.java
import java.util.*;

class DnDCharacter {

    private int _ability = -1;
    int ability() {
        if (_ability == -1)
            _ability = getRandomThrows();
        return _ability;

    }


    int modifier(int input) {
        return (int)Math.floor(((input-10.0)/2));
        //return (int)Math.floor(((this.getConstitution()-10.0)/2));
    }

    int _strength = -1;
    int getStrength() {
        if (_ability == -1)
            _ability = getRandomThrows();
        return _ability;
    }

    int _dexterity = -1;
    int getDexterity() {
        if (_dexterity == -1)
            _dexterity = getRandomThrows();
        return _dexterity;
    }

    int _constitution = -1;
    int getConstitution() {
        if (_constitution == -1)
            _constitution = getRandomThrows();
        return _constitution;
    }

    int _intelligence = -1;
    int getIntelligence() {
        if (_intelligence == -1)
            _intelligence = getRandomThrows();
        return _intelligence;
    }

    int _wisdom = -1;
    int getWisdom() {
        if (_wisdom == -1)
            _wisdom = getRandomThrows();
        return _wisdom;
    }

    int _charisma = -1;
    int getCharisma() {
        if (_charisma == -1)
            _charisma = getRandomThrows();
        return _charisma;
    }

    int _hitpoints = -1;
    int getHitpoints() {
        if (_hitpoints == -1)
            _hitpoints = modifier(getConstitution())+10;
        return _hitpoints;
    }

    int getRandomThrows()
    {
        int lowestIndex= 0;
        ArrayList<Integer> rollList = new ArrayList<Integer>();
        for (int i=0; i < 4; i++)
            rollList.add((int)Math.ceil((Math.random()*6)));
        for (int i=0; i<rollList.size(); i++)
            if (rollList.get(lowestIndex) > rollList.get(i))
                lowestIndex = i;
        rollList.remove(lowestIndex);
        int newTotal = 0;
        for (Integer data: rollList)
            newTotal += data;
        return newTotal;
    }
}
